ast.js

```diff
@@ -1,52 +1,109 @@
+const parse = require('./build/pegjs-parser').parse;
+const clues = require('clues');
+const MAX_DEPTH = 50;
+const INDENTATION_AT_DEPTH = [...new Array(MAX_DEPTH)].map((d,i) => '  '.repeat(i));
+
+function pathParser(path) {
+  try {
+    return parse(path.trim());
+  }
+  catch (e) {
+    throw e.message; // make sure it isn't `Error`
+  }
+}
+
 function astToCluesPath(node) {
   return astToString(node.paren || node);
 }
 
-function astToString(node) {
+function astToString(node, pretty=false, depth=0, ignoreFirstIndent=false) {
+  if (depth > MAX_DEPTH) {
+    depth = MAX_DEPTH;
+  }
+
+  let result;
   if (node.piped) {
-    return node.piped.map(n => astToString(n)).join('|');
+    let prefix = '';
+    if (pretty && node.piped.length > 1) {
+      prefix = `\n${INDENTATION_AT_DEPTH[depth]}`;
+    }
+    result = node.piped.map((n,i) => ((!pretty || ignoreFirstIndent && i === 0) ? '' : prefix) + astToString(n, pretty, depth)).join(pretty ? ', ' : ',');
+  }
+  else if (node.equation) {
+    result = `${astToString(node.equation.left, pretty, depth)}${node.operation}${astToString(node.equation.right, pretty, depth)}`;
+  }
+  else if (node.paren) {
+    result = `(${astToString(node.paren, pretty, depth, true)})`;
   }
-  if (node.equation) {
-    return `${astToString(node.equation.left)}${node.operation}${astToString(node.equation.right)}`;
+  else if (node.remoteLink) {
+    result = `\${${astToString(node.remoteLink, pretty, depth+1, true)}}`;
```

Comment:
should this be depth+1?

Reply:
The pretty printing isn't perfect -- and as long as it gets parsed correctly on the other side it doesn't matter THAT much. 

but I think I did this on purpose.

```
property.
  residential.
    transactions.
      all.
        select.
          amount<add(
            5, 
            ${global_app.
                data.
                  APPRAISAL}
          )
```
vs

```
property.
  residential.
    transactions.
      all.
        select.
          amount<add(
            5, 
            ${global_app.
              data.
                APPRAISAL}
          )
```

Reply:
you can see the `${` makes things a little weird with indentation